Add RPC call abandontransaction

morcos opened the pull request:
Unconfirmed transactions that are not in your mempool either due to eviction or other means may be unlikely to be mined.  abandontransaction gives the wallet a way to no longer consider as spent the coins that are inputs to such a transaction.  All dependent transactions in the wallet will also be marked as abandoned.

@laanwj also for 0.12

This is the basic functionality.  
There are more things to add though.  I have an RPC test  in progress, but if anyone else wants to work on the remaining items, please do:
- Return abandoned status in listtransactions
- Return abandoned status in GUI
- Fix any issues with how abandoned txs should sort
- Add a way to abandon transactions from GUI

I built this off of #7306 to make sure the tests would work correctly.

luke-jr commented:
This is way too late for 0.12...

morcos commented:
@luke-jr I don't think we can release 0.12 without this.  With mempool eviction combined with the new behaviour that unconfirmed txs still tie up the coins they spend even if the txs are not in the mempool, users may end up with permanently unspendable coins frequently.

We discussed making this change in conjunction with the other changes back in November, but it appears to have slipped through the cracks.

luke-jr commented:
This situation isn't really any worse than the status quo from older versions...

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -455,8 +457,10 @@ bool CWallet::IsSpent(const uint256& hash, unsigned int n) const
-        if (mit != mapWallet.end() && mit->second.GetDepthInMainChain() >= 0)
-            return true; // Spent
+        if (mit != mapWallet.end()) {
+            if (mit->second.GetDepthInMainChain() > 0  || (mit->second.GetDepthInMainChain() == 0 && !mit->second.isAbandoned()))
```
`CMerkleTx::GetDepthInMainChain()` is slightly expansive (and `IsSpent()` gets called a lot). Would caching `mit->second.GetDepthInMainChain()` make sense? Or do we expect the compiler does it internally (I don't think so)?

morcos replied:
sure can fix

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -48,6 +48,8 @@ bool fSendFreeTransactions = DEFAULT_SEND_FREE_TRANSACTIONS;
  */
 CFeeRate CWallet::minTxFee = CFeeRate(DEFAULT_TRANSACTION_MINFEE);
 
+const uint256 CMerkleTx::abandonHash(uint256S("0000000000000000000000000000000000000000000000000000000000000001"));
```
nit: I probably would do it also that way,... but, maybe a cleaner way would be to use `CWalletTx::mapValue` (a flexible kv store) which also would not require a db-scheme update.

morcos replied:
I debated doing that and maybe it would have been better, but I liked the way updating the hashBlock automatically made the transaction no longer abandoned with this approach.

jonasschnelli replied:
> I debated doing that and maybe it would have been better, but I liked the way updating the hashBlock automatically made the transaction no longer abandoned with this approach.

fair enough.

jonasschnelli commented:
Concept ACK.
I agree it late for 0.12,... but I don't take a position on this (if it's saver to go with or without a such feature).

Would be nice to see some tests (especially when we tag this for 0.12).

morcos committed:
Make wallet descendant searching more efficient

morcos committed:
Add new rpc call: abandontransaction

Unconfirmed transactions that are not in your mempool either due to eviction or other means may be unlikely to be mined.  abandontransaction gives the wallet a way to no longer consider as spent the coins that are inputs to such a transaction.  All dependent transactions in the wallet will also be marked as abandoned.

morcos commented:
Addressed @jonasschnelli's comment (thanks for quick review!) and reordered commits for simpler reading.

morcos commented:
oops, i didn't mean to push that RPC test yet, it was still in progress, but might as well leave it there for now...

kanzure commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1764,6 +1764,40 @@ UniValue gettransaction(const UniValue& params, bool fHelp)
+        throw runtime_error(
+            "abandontransaction \"txid\"\n"
+            "\nMark in-wallet transaction <txid> as abandoned\n"
+            "This will mark this tx and all its in-wallet descendants as abandoned which will allow\n"
```
consistency nit: all the other fHelp messages use "transaction" and "transactions" instead of "tx" and "txs".

kanzure commented:
I think a name like "abandonlocaltransaction" or "abandonunconfirmedtransaction" could help to minimize ambiguity.  But I would consider this a very low priority suggestion, because this RPC call is only available when the wallet is enabled.

MarcoFalke commented:
> abandonunconfirmedtransaction

How would you type that without any typos? I'd rather make the rpc short ("abandontx") and the documentation/GUI verbose.

MarcoFalke commented:
Concept ACK  6d74a63154cca756d698d6022c694b21e7f43ac5

morcos committed:
Add RPC test for abandoned and conflicted transactions.

morcos commented:
Thanks for the review. 
Addressed nits in RPC help and finalized the RPC test

MarcoFalke commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -105,6 +105,7 @@
     'prioritise_transaction.py',
     'invalidblockrequest.py',
     'invalidtxrequest.py',
+    'abandonconflicts.py',
```
abandonconflicts.py: No such file or directory

MarcoFalke commented on qa/rpc-tests/abandonconflict.py:
```diff
@@ -0,0 +1,153 @@
+
+    def setup_network(self):
+        self.nodes = []
+        self.nodes.append(start_node(0, self.options.tmpdir, ["-debug","-logtimemicros","-minrelaytxfee=0.00001"], binary="bitcoind"))
```
Is the `"-debug","-logtimemicros"` part required?

sdaftuar replied:
I think we should drop the `binary="bitcoind"` part -- by default `start_node` can pick up the binary from the BITCOIND environment variable which is preferred when set.

MarcoFalke commented on qa/rpc-tests/abandonconflict.py:
```diff
@@ -0,0 +1,153 @@
+
+        sync_blocks(self.nodes)
+        newbalance = self.nodes[0].getbalance()
+        assert(balance - newbalance < Decimal("0.001")) #no more than fees lost
```
Nit: `Decimal("0.00003")` should be enough.

MarcoFalke commented:
Review tests ACK 21702ff

@morcos Please fix the travis issue (file not found)

morcos commented:
oops sorry.
fixed typo and removed bitcoind argument from rpc test.

MarcoFalke commented on qa/rpc-tests/abandonconflict.py:
```diff
@@ -0,0 +1,153 @@
+        balance = newbalance
+
+        # Send child tx again so its unabandoned
+        self.nodes[0].sendrawtransaction(signed2["hex"])
```
Shouldn't `resendwallettransactions` do the same? I get the empty array on `resendwallettransactions`.

morcos replied:
I thought it made sense that abandoned transactions would not be automatically resent.
That way when you restart a bitcoind with transactions that you've tried to abandon, it wouldn't automatically try to send them all again

MarcoFalke commented:
QT-tested ACK d716866

![screenshot from 2016-01-08 20-41-15](https://cloud.githubusercontent.com/assets/6399679/12207715/839120a8-b648-11e5-9ee4-a74eb4ebdfe8.png)
Nit: Qt balance != getwalletinfo

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -768,6 +779,63 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransaction& tx, const CBlock* pbl
+            wtx.nIndex = -1;
+            wtx.setAbandoned();
+            wtx.MarkDirty();
+            wtx.WriteToDisk(&walletdb);
```
I think here we also need a `NotifyTransactionChanged(this, wtx.GetHash(), CT_UPDATED);` Otherwise the signal listeners (UI) can't recalculate the balance.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -768,6 +779,63 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransaction& tx, const CBlock* pbl
+            BOOST_FOREACH(const CTxIn& txin, wtx.vin)
+            {
+                if (mapWallet.count(txin.prevout.hash))
+                    mapWallet[txin.prevout.hash].MarkDirty();
```
Probably also here:
`NotifyTransactionChanged(this, txin.prevout.hash, CT_UPDATED);`

jonasschnelli commented:
Tested ACK (also ack on a 0.12 bp, because it's a wallet only change)
Nits:
* missing update signal (see above)
* missing flag in `listtransactions`.

jonasschnelli committed:
[Wallet] Call notification signal when a transaction is abandoned

jonasschnelli commented:
Addressed the nits and a cosmetic change for GUI in: https://github.com/morcos/bitcoin/pull/6

morcos commented:
@jonasschnelli , thanks!  I added your notify commit.  I left the other changes for a separate pull.

morcos commented:
Ok updated after speaking to @jonasschnelli 
I think this is good now

PRabahy commented:
Does https://github.com/bitcoin/bitcoin/blob/master/contrib/bitcoind.bash-completion need to get updated with this change?

MarcoFalke commented:
@PRabahy sure, but I don't think this is part of this pull.

laanwj commented on src/wallet/wallet.h:
```diff
@@ -155,6 +155,9 @@ struct COutputEntry
 class CMerkleTx : public CTransaction
 {
+private:
+    static const uint256 abandonHash;
```
Can you explain why this static field on CMerkleTx is necessary? (and add a comment) What is it used for, what coordinates/synchronizes access to it?
(e.g. why is there only one "abandoned transaction hash"? why is it not passed in some other way?)

morcos commented:
Addressed @laanwj's concern about clearly identifying constant
